flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java

```diff
@@ -42,21 +42,22 @@
   private static final long serialVersionUID = 1L;
 
   private final TableLoader tableLoader;
-  private final Schema tableSchema;
   private final FileIO io;
   private final EncryptionManager encryption;
   private final ScanContext context;
+  private final RowDataFileReader rowDataReader;
 
-  private transient RowDataIterator iterator;
+  private transient DataIterator<RowData> iterator;
   private transient long currentReadCount = 0L;
 
   FlinkInputFormat(TableLoader tableLoader, Schema tableSchema, FileIO io, EncryptionManager encryption,
                    ScanContext context) {
     this.tableLoader = tableLoader;
-    this.tableSchema = tableSchema;
     this.io = io;
     this.encryption = encryption;
     this.context = context;
+    this.rowDataReader = new RowDataFileReader(tableSchema,
```

Comment:
Is there necessary to expose the `RowDataFileReader` to the FlinkInputFormat ?  The calling chain is actually:  FlinkInputFormat ( whole scan) ->  DataIterator (CombinedScanTask reader) -> RowDataFileReader ( FileScanTask reader ).   I mean the FlinkInputFormat won't call the FileScanTask reader directly,  so  I'd prefer to hidden the internal details about how to construct the RowDataFileReader inside DataIterator. This is more in line with the interface design of software design.

Reply:
`DataIterator` is a generic type. The actual `FileScanTaskReader` needs to be provided when constructing `DataIterator`.  Previously, `FlinkInputFormat` constructs the extended class `RowDataIterator`, which is removed in the PR. Now `FlinkInputFormat` constructs the `RowDataFileScanTaskReader` and pass it into `DataIterator` constructor for composition.

Reply:
You mean there will be other data type applied to the generic `DataIterator` ( For example for flip-24 ) ?  There seems no other `FileScanTaskReader` implementation except the RowDataFileScanTaskReader.  I think we may don't want to the generic type for the `DataIterator`,  then we also don't need the extra `FileScanTaskReader` (we could just rename the RowDataFileScanTaskReader to FileScanTaskReader directly), and finally we construct the FileScanTaskReader inside the DataIterator.   How do you think ?

Reply:
It is correct that in iceberg-flink, there is no other impl of `FileScanTaskReader` other than `RowDataFileScanTaskReader`. But At Netflix and Apple, we use Avro schema for the Kafka data and deserialize bytes into Avro Record. For Iceberg source, we also want to deserialize data into Avro Record. Hence we will plug in `AvroGenericRecordFileScanTaskReader`. That is one of the motivation of this refactoring.

Reply:
That make sense for providing an abstracted `FIleScanTaskReader` in the offical apache iceberg repo,  what concerns me is:  we may get this interface refactored  or removed when other developers propose a new pull request ,  if we don't have other impl in apache repo.  Will we have other impl for the flip-27 work in offical repo ?

Reply:
@openinx I extracted `AbstractFileScanTaskReader` as you suggested. 

In flip-27 source in apache iceberg repo, we won't have other `FileScanTaskReader` impl.

Reply:
Sorry, I think I did not describe this more clear.  I mean the newly introduced `FileScanTaskReader` interface is nice to have in apache repo, so that others could provides  their own implementation, such as `AvroGenericRecordFileScanTaskReader`.   I don't think the latest commit's `AbstractFileScanTaskReader`  will help to share common code between the `AvroGenericRecordFileScanTaskReader`  & `RowDataFileScanTaskReader`  because its exposed classes such as `DeleteFilter`, `newAvroIterator`, `InputFilesDecryptor` are not stable enough, we may change/refactor them in the following PRs (The DeleteFilter will need a big refactor for supporting unified v2 compaction).   

Let's just revert the commit https://github.com/apache/iceberg/pull/2905/commits/72348c7898c42566a8a5a071426b32f8ea746cfd,  all the other things look great to me now,  I plan to merge this PR.  Thanks for the great work @stevenzwu  !

Reply:
@openinx sorry for the misunderstanding. I have reverted the last commit by hard reset and force push